### PR TITLE
Multi-version configuration initial support

### PR DESCRIPTION
**Pull request by jasonwzm**
Provides a permanent fix for https://github.com/istio/istio/issues/20449

Initial implementation of https://github.com/istio/istio/issues/2119

This PR adds multi-version config for Istio resources. As a result, validation webhook is able to validate v1beta1 and v1alpha3 networking configuration (see changes in `validation.go`).

Version conversion coming in another PR.

[X] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by howardjohn** [comment]
Please revert the assetts.gen change, we do not check that in anymore

**Review by howardjohn — COMMENTED** [review]
lgtm overall

**Comment by howardjohn** [comment]
> Version conversion coming in another PR.

Can you please elaborate on this? Is this https://docs.google.com/document/d/1js1x0vf9H7kqknGwTyHhzPe9L7gG-dNIhC6ypPsKm3g?

**Comment by jasonwzm** [comment]
@howardjohn yes, I think the main concern you had was around multi control plane. When the doc was written, we did not think too much about that. Let me sync offline on this topic and will add to the doc and also discuss with you. This PR should be good to go once tests pass.

**Review by esnible — APPROVED** [review]

**Comment by howardjohn** [comment]
@jasonwzm why are you sending a PR for a design that is not approved?

**Comment by jasonwzm** [comment]
@howardjohn This PR is more about a permanent fix for https://github.com/istio/istio/issues/20449 rather a complete implementation of the multi version conversion design. This PR does not involve any version conversion.

**Comment by howardjohn** [comment]
@jasonwzm I don't have concerns with this PR, just the theoretical future pr

**Comment by howardjohn** [comment]
I was waiting for tests to pass to review

**Comment by jasonwzm** [comment]
@howardjohn will talk more about your concerns in the doc. fixing the unit test now.

**Review by howardjohn — CHANGES_REQUESTED** [review]
I am really confused here. Why are we watching multiple versions of the same object? This just means 2x the load on the api server for no benefit. As far as I understand it, the whole basis for k8s versioning is that you just read/write a single version and it automagically (through conversion webhooks or otherwise) handles things for you?

At the very least, we are triggering 2x events on every CRD change for things with multiple versions, so we cannot merge as is

File: galley/pkg/config/source/kube/apiserver/source.go
Comment (howardjohn) [inline galley/pkg/config/source/kube/apiserver/source.go:159]: what happens when an MCP client requests a different version?

**Review by howardjohn — CHANGES_REQUESTED** [review]
This does not fix the main problem - Pilot CRD reader in kube/crd/controller. Nothing is really using MCP anymore

**Comment by dex80526** [comment]
@jasonwzm: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by howardjohn** [comment]
Is this still valid?

**Pull request closed without merging**
